src/analysis/scoping_analysis.cpp

```diff
@@ -401,11 +397,21 @@ class NameCollectorVisitor : public ASTVisitor {
         }
     }
 
-    bool visit_generatorexp(AST_GeneratorExp* node) override {
+    // helper methods for visit_{generatorexp,dictcomp,setcomp}
+    void visit_comp_values(AST_GeneratorExp* node) { node->elt->accept(this); }
+    void visit_comp_values(AST_SetComp* node) { node->elt->accept(this); }
+    void visit_comp_values(AST_DictComp* node) {
+        node->key->accept(this);
+        node->value->accept(this);
+    }
+
+    template <typename CompType> bool visit_comp(CompType* node) {
+        // NB. GeneratorExps evaluate their first for-subject's expression outside of the function scope they create.
+        // dictcomps & setcomps don't.
```

Comment:
Hmm...

```python
x = "before"
{i for i in [x for x in xrange(5)]}
print x
```
CPython prints 4, this PR prints "before".  It's a lot more esoteric than the generatorexp version though (with generator expressions it affects not just scoping but the order in which things get evaluated).